feat(components/datetime): convert a user-entered digit into a date for the datepicker component

Blackbaud-TrevorBurch opened the pull request.

Blackbaud-SteveBrush commented on libs/components/datetime/src/lib/modules/datepicker/datepicker-input.directive.ts:
```diff
@@ -534,7 +576,7 @@ export class SkyDatepickerInputDirective
     }
 
-    if (dateValue && isValidDateString) {
+    if (dateValue && dateValue instanceof Date && isValidDateString) {
```
The variable name `dateValue` was written to imply that the value was an instance of `Date`.  Is there a way to keep the variable of type `Date` to avoid having the `instanceof Date` checks in multiple places?

Blackbaud-TrevorBurch replied:
I changed this to `convertedDate` and added a comment.

We can't assume it is always a `Date` anymore as we don't always convert it if it is a string that is not a valid date or one that is numeric but outside the current month's number of days.

Blackbaud-SteveBrush commented on libs/components/datetime/src/lib/modules/datepicker/datepicker-input.directive.ts:
```diff
@@ -353,16 +353,57 @@ export class SkyDatepickerInputDirective
     }
 
-    const value: any = control.value;
+    const value: Date | string = control.value;
```
Isn't the return value of the control technically an `any`? Number, string, undefined, null, Date, etc.?

Blackbaud-TrevorBurch replied:
Good catch. Switched this to `unkown`.

Blackbaud-SteveBrush commented on libs/components/datetime/src/lib/modules/datepicker/datepicker-input.directive.ts:
```diff
@@ -503,11 +527,29 @@ export class SkyDatepickerInputDirective
+        } else {
+          dateValue = value;
+        }
+      } else {
```
If we could bake this logic into a self-describing private method, that would help for readability.

Blackbaud-TrevorBurch replied:
I refactored code to try to make the flow easier to understand and have it be more concise.

Blackbaud-SteveBrush commented on libs/components/datetime/src/lib/modules/datepicker/datepicker-input.directive.ts:
```diff
@@ -353,16 +353,57 @@ export class SkyDatepickerInputDirective
+      const isDateValid = this.dateFormatter.dateIsValid(value);
+
+      /* sanity check */
+      /* istanbul ignore if */
```
We're ignoring the `if` statement for coverage?

Blackbaud-TrevorBurch replied:
Good call - I did that early on. I've added tests to cover this and removed the ignore statement here.

Blackbaud-SteveBrush commented on libs/components/datetime/src/lib/modules/datepicker/datepicker-input.directive.ts:
```diff
@@ -440,22 +461,46 @@ export class SkyDatepickerInputDirective
+  /**
+   * Gets the date value from a value - if possible. Will not convert unconvertable dates or numbers outside of the current month's number of days
+   */
+  private getDateValue(value: unknown): unknown {
```
The name of the method implies that I'm getting a `Date` value. What if we made the return type `Date | null` so the value is more explicit and easier to work with. The logic you wrote below will still work with that return type as well.
```
private getDateValue(value: unknown): Date | null {}

private getDateFromShortcut(value: string): Date | null {}

private updateValue(value: unknown, emitEvent = true): void {
  const dateValue = this.getDateValue(value);

  // Your existing `if` checks, here.
}
```

Blackbaud-TrevorBurch replied:
I've refactored things just a tad to now allow for just `Date | undefined` as the return type from these methods. And then the existing logic handles if `undefined` is returned. Sorry that I had overcomplicated things - it took me a bit after you commented this to even be able to step back and see this path of logic.

Blackbaud-SteveBrush reviewed (commented):
If we could start the commit message with a verb, that would be ideal. Something to the effect of, "convert a user-entered digit into a date for the datepicker component"?

Blackbaud-TrevorBurch committed:
feat(components/datetime): convert a user-entered digit into a date for the datepicker component

Datepicker numeric input is now translated to a date in the current month if the input is within the
current month's number of days.

BREAKING CHANGE: Datepicker numeric input is now translated to a date in the current month if the
input is within the current month's number of days. Numeric input outside of the current month's
number of days is now treated as invalid and is not converted to a `Date` object.
